### PR TITLE
ROS 2: Library with dynamic modes API

### PR DESCRIPTION
**Pull request by bkueng**
### Goal
Create a library and interface for writing ROS2 applications that control the drone. This includes high-level navigation tasks all the way down to direct actuator controls.

It goes much further than the existing offboard mode:
- the ROS2 mode behaves the same as an FMU-internal mode:
  - it can provide a name that is selectable in the GCS
  - it can provide its own arming checks and mode requirements
  - integrated into failsafe state machine
- there are clear rules for who is controlling the drone (see mode executor below)
- it's well-defined which setpoints can be sent
- a ROS2 mode can replace an FMU-internal mode
- multiple ROS2 modes & executors can co-exist and run (up to 8 can be registered)

### Concepts
This adds a few new concepts based on existing ones:
image

**Flight Mode**
- Is passive, does not execute other modes (or commands)
- Is activated via a mode executor / user / FC
- Selects and sends setpoints to the FC (e.g. velocity)
- Has a name displayed by the GCS
- Can configure its mode requirements

**Mode Executor**
- Can select modes and send commands when it is in charge (and wait for completion)
- Selects exactly one ‘owned’ mode. This is typically a custom mode that then generates the setpoints to perform the desired task.
- The executor is in charge when the owned mode gets selected and stays in charge until failsafe is triggered or the user switches modes.
- There can be multiple executors, but at most one is in charge at any given time
- In terms of FMU implementation, the `navigator` module corresponds to the mode executor (it currently does more than that though, also implementing several modes)
- Note that this is not a recursive definition (executors cannot activate other executors)

**Config Overrides**
Both, modes and executors can define configuration overrides. Two are defined and implemented currently:
- disable auto-disarm. Allows for landing and then taking off again
- ability to defer non-essential failsafes. Allows for e.g. an executor to execute a critical action w/o being interrupted by low-battery failsafe.

**Mode Replacement**
An external mode can replace an internal one, for example a sophisticated RTL. Both triggering RTL failsafe or a user selecting RTL in the GCS will then use the external mode instead. If the external mode crashes, the internal one will be used as fallback. This is currently based on regular arming check request timeout (could later be extended to include setpoint timeouts too).

**Mode Completion**
For an executor to know when to continue, there's a new topic `mode_completed` added, published by takeoff, rtl, etc.

### Example Usage
A flight mode looks like this:
```cpp
class FlightModeTest : public ModeBase
{
public:
	FlightModeTest(rclcpp::Node &node)
		: ModeBase(node, Settings{"My Mode"}, ModeRequirements::manualControlledPosition())
	{
		setSetpointUpdateRate(60.f);
	}

	void checkArmingAndRunConditions(HealthAndArmingCheckReporter &reporter) override
	{
		// custom checks
	}

	void onActivate() override
	{
		setpoints().configureSetpointsSync({});
	}

	void onDeactivate() override {}

	void updateSetpoint() override
	{
		// send setpoint
	}

private:
};
```

And a mode executor in the form of a simple state machine:
```cpp
class ModeExecutorTest : public ModeExecutorBase
{
public:
	ModeExecutorTest(rclcpp::Node &node, ModeBase &owned_mode)
		: ModeExecutorBase(node, ModeExecutorBase::Settings{false}, owned_mode),
		  _node(node)
	{}

	enum class State {
		Reset,
		WaitForArming,
		Arming,
		TakingOff,
		MyMode,
		RTL,
		WaitUntilDisarmed,
		Shutdown,
	};

	void onActivate() override
	{
		runState(State::WaitForArming, Result::Success);
	}

	void onDeactivate(DeactivateReason reason) override
	{}

	void runState(State state, Result previous_result)
	{
		if (previous_result != Result::Success) {
			RCLCPP_ERROR(_node.get_logger(), "State %i: previous state failed: %s", (int)state, resultToString(previous_result));
			return;
		}

		RCLCPP_DEBUG(_node.get_logger(), "Executing state %i", (int)state);

		switch (state) {
		case State::Reset:
			break;

		case State::WaitForArming:
			waitReadyToArm([this](Result result) { runState(State::Arming, result); });
			break;

		case State::Arming:
			arm([this](Result result) { runState(State::TakingOff, result); });
			break;

		case State::TakingOff:
			takeoff([this](Result result) { runState(State::MyMode, result); });
			break;

		case State::MyMode:
			scheduleMode(ownedMode().id(), [this](Result result) { runState(State::RTL, result); });
			break;

		case State::RTL:
			rtl([this](Result result) { runState(State::WaitUntilDisarmed, result); });
			break;

		case State::WaitUntilDisarmed:
			waitUntilDisarmed([this](Result result) { runState(State::Shutdown, result); });
			break;

		case State::Shutdown:
			rclcpp::shutdown();
			break;
		}
	}

private:
	rclcpp::Node &_node;
};
```

### Other
- Implements and depends on https://github.com/mavlink/mavlink/pull/1915
- I have some example modes if anyone is interested
- Flash usage is ~20KB, and ~8KB for flash constrained builds

### TODO's (for later)
- main missing part: control interface (defining setpoints)
- test on hw
- param handling
- move the lib into own repo, add `px4_ros_com` (the transformations)
- metadata handling for custom checks
- CI: run tests
- mode activation by command with configuration (e.g. orbit)
- log mode names
- ...

### CODE REVIEW

**Comment by Jaeyoung-Lim** [comment]
This is something that was really needed for enabling more higher level autonomy in PX4

**Comment by Jaeyoung-Lim** [comment]
@bkueng I think this is probably no longer compatible with the [px4_msgs](https://github.com/PX4/px4_msgs) message definitions

what would be minimum work package left so that we can get this incorporated in main?

**Comment by TOTON95** [comment]
@bkueng I was checking the CI checks, and most have a problem with a particular line: `vehicle_visual_odometry = 256`, it is an *uint8_t* variable, and making it equal to 256 pushes it out of range.

**Comment by TOTON95** [comment]
(For the record of what happened at the Discord Server) At the [Q&A Session](https://discuss.px4.io/t/px4-community-q-a-august-02-2023/33434/1), @dagar suggested me to change the `ORB_ID` to  `uint16_t`, which indeed solves the compilation for this WIP. However, it produced a file which is larger for its target (pixhawk 6c). Later, he encouraged me to define my needs using this API before we continue with further experiments.

I'll update about it once I create a ROS2-PX4 setup for this purpose. 


image

**Comment by bkueng** [comment]
@TOTON95 I applied the uorb id fix, let me know if you want to test. Currently it still involves a few manual steps go get everything working.

**Comment by TOTON95** [comment]
> @TOTON95 I applied the uorb id fix, let me know if you want to test. Currently it still involves a few manual steps go get everything working.

Thank you, I would like to test the latest changes that you did. Could you indicate to me what manual steps I need to perform for these tests?

**Comment by TOTON95** [comment]
image

So far is looking good. Still, I need to test it in some experiments though. Due to my job, I haven't been able to design some for testing, I will attempt to create a few for this weekend.

**Comment by bkueng** [comment]
@TOTON95 these are the steps required for testing:
- You'll need the `modes` branch from px4_msgs: https://github.com/PX4/px4_msgs/tree/modes
- add this to your ros workspace: https://github.com/PX4/px4_sdk
- then you can implement your own mode, check out the examples: https://github.com/PX4/px4_sdk/tree/main/examples/cpp/modes
- then run sitl as normal, run the xrce agent, and start the ros mode.
- to activate the mode, there's upcoming QGC changes. For now you can use `commander mode ext1` on the px4 shell.

**Comment by TOTON95** [comment]
> @TOTON95 these are the steps required for testing:
> 
>     * You'll need the `modes` branch from px4_msgs: https://github.com/PX4/px4_msgs/tree/modes
> 
>     * add this to your ros workspace: https://github.com/PX4/px4_sdk
> 
>     * then you can implement your own mode, check out the examples: https://github.com/PX4/px4_sdk/tree/main/examples/cpp/modes
> 
>     * then run sitl as normal, run the xrce agent, and start the ros mode.
> 
>     * to activate the mode, there's upcoming QGC changes. For now you can use `commander mode ext1` on the px4 shell.

Thanks for your reply, I will try to run these instructions this weekend.

File: src/modules/mavlink/streams/CURRENT_MODE.hpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+	bool send() override
+	{
+		vehicle_status_s vehicle_status;
+
+		if (_vehicle_status_sub.update(&vehicle_status)) {
+			mavlink_current_mode_t current_mode{};
+			current_mode.custom_mode = get_px4_custom_mode(vehicle_status.nav_state).data;
+			current_mode.intended_custom_mode = get_px4_custom_mode(vehicle_status.nav_state_user_intention).data;
+			current_mode.standard_mode = (uint8_t) mode_util::getStandardModeFromNavState(vehicle_status.nav_state);
```
Comment (dagar) [inline src/modules/mavlink/streams/CURRENT_MODE.hpp:69]: Move getStandardModeFromNavState() to the mavlink module or a library?

**Comment by dagar** [comment]
Why would you need to apply overrides for a mode executor rather than just the mode?
 
Other than that why does commander even need to know about the existence of ModeExecutors?

File: src/modules/mavlink/streams/CURRENT_MODE.hpp (hunk 1 above)
Comment (bkueng) [reply src/modules/mavlink/streams/CURRENT_MODE.hpp:69]: I can do that.

**Comment by bkueng** [comment]
> Why would you need to apply overrides for a mode executor rather than just the mode?

For example to disable auto-disarm while landing.

> Other than that why does commander even need to know about the existence of ModeExecutors?

Mainly as single source of truth and to tell the rest of the system which one is in charge. Let me know if you have an equivalent alternative.

**Comment by dagar** [comment]
> > Why would you need to apply overrides for a mode executor rather than just the mode?
> 
> For example to disable auto-disarm while landing.

That seems like a good example for a particular mode, but why would someone need it to apply to every mode in a mode executor?

**Comment by bkueng** [comment]
>      That seems like a good example for a particular mode, but why would someone need it to apply to every mode in a mode executor?

When triggering the landing from the mode executor.

File: msg/RegisterExtComponentRequest.msg [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+uint16 LATEST_PX4_SDK_API_VERSION = 1 # API version compatibility. Increase this on a breaking semantic change. Changes to any message field are detected separately and do not require an API version change.
+
+uint16 px4_sdk_api_version   # Set to LATEST_PX4_SDK_API_VERSION
+
+# Components to be registered
+bool register_arming_check
+bool register_mode                 # registering a mode also requires arming_check to be set
+bool register_mode_executor        # registering an executor also requires a mode to be registered (which is the owned mode by the executor)
+
+bool enable_replace_internal_mode  # set to true if an internal mode should be replaced
```
[the remark this answers is not in the document]
Comment (bkueng) [reply msg/RegisterExtComponentRequest.msg:16]: Not necessarily.
Those details are good reasons to me for decoupling and making it explicit.

[the remark this answers is not in the document]
Comment (dagar) [reply msg/RegisterExtComponentRequest.msg:16]: What I'm getting at is avoiding the potential confusion of replacing an internal mode with an offboard mode of a different name.

[the remark this answers is not in the document]
Comment (bkueng) [reply msg/RegisterExtComponentRequest.msg:16]: Yes it puts some responsibility on the developer. The same holds for the actual behavior of the mode.

**Comment by fury1895** [comment]
@bkueng While testing with this branch I managed to end up in the situation, where the only external mode was enumerated as external mode 2 instead of 1.
I think, what happened is that we had already 2 different external modes interfacing with this flight controller, and with removing one of them, the enumeration did not get reset.
Is this behavior something you'd expect and can easily be fixed?

**Comment by bkueng** [comment]
@fury1895 yes that is expected, to keep the assignment consistent. You can reset the `COM_MODE${i}_HASH` parameters to clear that.

**Comment by bkueng** [comment]
Rebased & CI is finally passing

**Comment by henrykotze** [comment]
During your PX4 talk, it sounded  like we are going to replace the internal navigator module with one of these ROS2 applications? Just want to get your thoughts on this? 

Context:
I  think I understood @dagar mentioning to me that we want all the modules to run as ROS2 nodes which allows the same code to be used internally or externally from the FMU. Is this a future look ahead of how for example the navigator module can be executed internally and externally, and its just based on hardware you are using.

**Review by tstastny — APPROVED** [review]

**Comment by bkueng** [comment]
> During your PX4 talk, it sounded like we are going to replace the internal navigator module with one of these ROS2 applications? Just want to get your thoughts on this?

No, at least not anytime soon. Long-term, when companion computers are standard and commonly used, I think it makes sense to abandon mission mode on fmu, and only use something on the higher-level compute side. This will probably be more use-case specific modes/apps instead of a single one that tries to fit everything (and works poorly for each one).

> I think I understood @dagar mentioning to me that we want all the modules to run as ROS2 nodes which allows the same code to be used internally or externally from the FMU. Is this a future look ahead of how for example the navigator module can be executed internally and externally, and its just based on hardware you are using.

Personally, I don't see much use of running something internally or externally. The low-level embedded FMU is just a different world compared to a full Linux system, with different limitations and requirements. Among the reasons to run on Linux is to make use of high-level features and more compute. And as soon as you start developping something on the Linux side you very quickly end up in a state where you cannot run it anymore on the more constrained RTOS environment.

**Pull request merged**